## **Maximizing Revenue with Data: 2019 Sales Trends and Product Performance**

### **Business Understanding**

#### **Problem Satement**

A client assigned by getINNOtized, has collected transactional data for the year 2019 but hasn't been able to effectively use this data to improve sales or operational efficiency. They need insights into sales performance, seasonal trends, product popularity, and city-level sales to help drive more sales and streamline operations.

#### **Goal and Objectives**

The main goal is to deliver a comprehensive business intelligence (BI) solution that helps the client:

- Identify trends and seasonality in sales.
- Analyze product performance to discover best- and worst-selling items.
- Compare sales across different time periods (monthly, weekly) for actionable insights.
- Analyze geographical sales performance to identify cities with higher demand.
- Segment products based on price and analyze their contribution to total sales.


#### **Stakeholders**
- Primary Stakeholders: Management team looking for sales and operational insights.
- Secondary Stakeholders: Sales and marketing teams who can use the insights for future campaigns.
- Analysts: Those responsible for deriving and communicating actionable insights.
- Operations Team: Can leverage insights for improving efficiency in delivering products to high-demand areas.

#### **Key Metrics and Success**
- Total Sales Revenue: Monthly and yearly revenue.
- Seasonality Metrics: Monthly/quarterly sales trends.
- Product Performance: Revenue and quantity sold by product.
- Geographic Metrics: Sales distribution by city.
- Product Category Performance: Revenue and quantity sold by product category (high-level vs. basic).
- Operational Efficiency: Timeliness of reporting, ease of extracting actionable insights, and ability to identify growth   opportunities

#### **Hypotheses**
Null Hypothesis (H0): There is no significant seasonality in sales across the year.

Alternate Hypothesis (H1): There is significant seasonality in sales, with certain months showing higher or lower sales trends

#### **Features**
1. Order ID: Identifies each sales transaction.
2. Product: Specifies the product sold.
3. Quantity Ordered: Represents the number of units ordered.
4. Price Each: Indicates the price per unit of the product.
5. Order Date: The date when the order was placed.
6. Purchase Address: The shipping or billing address associated with the order.

#### **Methodology** 
The project follows the CRISP-DM (Cross Industry Standard Process for Data Mining) methodology, structured as: 
- Business Understanding: Define objectives and requirements from the client to drive sales and improve efficiency
- Data Understanding: Collect and explore the dataset, which includes sales data for 2019 across various columns
- Data Preparation: Clean and transform the data, integrating sources from CSV files and a remote database
- Modelling: Apply appropriate data analysis techniques to uncover insights.
- Evaluation: Assess the results against business objectives to ensure they meet the client’s needs
- Deployment: Develop and deploy a Power BI dashboard to visualize the insights

#### **Analytical Questions**
1. How much money did we make this year? 

2. Can we identify any seasonality in the sales? 

3. What are our best and worst-selling products? 

4. How do sales compare to previous months or weeks? 

5. Which cities are our products delivered to most? 

6. How do product categories compare in revenue generated and quantities ordered? 

#### **Scope and Constraints**
Scope:
- Analysis will focus on the sales data for 2019, broken down by months, weeks, and product categories.
- Sales data for January to June will be extracted from CSV files, and data for July to December will be pulled from the database.
- The analysis will also involve comparison between high-level and basic products based on unit price thresholds.

Constraints:
- Data integration: Combining two different data sources (CSV and database).
- Time constraints for accessing, cleaning, and preparing the data.
- Potential inconsistencies in data formatting between the first and second halves of the year.


#### **Extra Information**
- The products with a unit price greater than $99.99 will be categorized as high-level products, and those below or equal to $99.99 will be considered basic products. This categorization will be critical for revenue comparisons across product types.
- To answer the questions efficiently, a blend of SQL for database querying and Python/Excel for data analysis and visualization will be required.
- This project must be completed in two weeks 






### **Data Understanding**

#### **Importation of libraries**

In [44]:
#Data manipulation and analysis
import pandas as pd
import numpy as np
from dotenv import dotenv_values
import os

# Managing environment variables
#from dotenv import dotenv_values


import warnings

warnings.filterwarnings('ignore')
 
# Database connectivity
import pyodbc
 
# Database ORM
from sqlalchemy import create_engine
 
# Data visualization
import matplotlib.pyplot as plt
import seaborn as sns
 

# Machine learning 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.preprocessing import LabelEncoder
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import OrdinalEncoder 
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import RobustScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import MaxAbsScaler
from sklearn.preprocessing import PowerTransformer
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV


#### **Data Loading**

In [45]:
#Loading Data from January to June
# List of CSV files (January to June)
csv_files = ['Sales_January_2019.csv', 'Sales_February_2019.csv', 'Sales_March_2019.csv', 
             'Sales_April_2019.csv', 'Sales_May_2019.csv', 'Sales_June_2019.csv']  

# Define the folder path
folder_path = r'C:\Users\USER\Desktop\Sales-Trends-2019\data'  

# Initialize an empty list to store DataFrames
df_firsthalf = []  

# Loop over the CSV files to load each into a DataFrame and append to list
for file in csv_files:
    # Correctly join the folder path and file name
    file_path = os.path.join(folder_path, file)
    
    # Read the CSV file into a DataFrame
    df = pd.read_csv(file_path)
    
    # Append the DataFrame to the list
    df_firsthalf.append(df)
    
    # Print the head of the DataFrame for each month for verification
    print(f"Head of {file.split('_')[1]} Data:")
    print(df.head())  
    print("\n")  # Add a line break for readability

# Now firsthalf_dataframes contains all the DataFrames for January to June


Head of January Data:
  Order ID                   Product Quantity Ordered Price Each  \
0   141234                    iPhone                1        700   
1   141235  Lightning Charging Cable                1      14.95   
2   141236          Wired Headphones                2      11.99   
3   141237          27in FHD Monitor                1     149.99   
4   141238          Wired Headphones                1      11.99   

       Order Date                       Purchase Address  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4  01/25/19 11:59          387 10th St, Austin, TX 73301  


Head of February Data:
  Order ID                   Product Quantity Ordered Price Each  \
0   150502                    iPhone                1        700   
1   150503     AA Batteries (4-pack)                1 

In [46]:
# Loading Data from July to December

# Define the connection string
connection_string = (
    'DRIVER={ODBC Driver 17 for SQL Server};'
    'SERVER=dap-projects-database.database.windows.net;'
    'DATABASE=dapDB;'
    'UID=capstone;'
    'PWD=Z7x@8pM$2w;'
)

# Connect to the SQL Server database
try:
    connection = pyodbc.connect(connection_string)
    print("Connection to dapDB successful!")
except Exception as e:
    print(f"Error connecting to the database: {e}")
    exit()

# List of table names 
table_names = ['Sales_July_2019', 'Sales_August_2019', 'Sales_September_2019',
               'Sales_October_2019', 'Sales_November_2019', 'Sales_December_2019']

# Create an empty list to store DataFrames
df_secondhalf = []

# Loop through each table and fetch data
for table in table_names:
    print(f"Fetching data from {table}...")
    query = f"SELECT * FROM {table}"  # SQL query to select all data from the table
    
    # Fetch the data into a Pandas DataFrame
    df = pd.read_sql(query, connection)
    
    # Display the first few rows 
    print(df.head())
    
    # Save the data to a CSV file 
    csv_filename = f"{table}.csv"
    df.to_csv(csv_filename, index=False)
    print(f"Data from {table} saved to {csv_filename}")
    
    # Append the DataFrame to the list
    df_secondhalf.append(df)



# Close the connection
connection.close()


Connection to dapDB successful!
Fetching data from Sales_July_2019...
   Order_ID                   Product  Quantity_Ordered  Price_Each  \
0  222910.0  Apple Airpods Headphones               1.0      150.00   
1  222911.0             Flatscreen TV               1.0      300.00   
2  222912.0     AA Batteries (4-pack)               1.0        3.84   
3  222913.0     AA Batteries (4-pack)               1.0        3.84   
4  222914.0    AAA Batteries (4-pack)               5.0        2.99   

           Order_Date                  Purchase_Address  
0 2026-07-19 16:51:00   389 South St, Atlanta, GA 30301  
1 2005-07-19 08:55:00     590 4th St, Seattle, WA 98101  
2 2029-07-19 12:41:00    861 Hill St, Atlanta, GA 30301  
3 2028-07-19 10:15:00   190 Ridge St, Atlanta, GA 30301  
4 2031-07-19 02:13:00  824 Forest St, Seattle, WA 98101  
Data from Sales_July_2019 saved to Sales_July_2019.csv
Fetching data from Sales_August_2019...
   Order_ID                     Product  Quantity_Ordered  P

#### **Insights**
- The datasets retrieved from the one drive are placed in the variable firsthalf and the datasets retrieved from the remote database are in the vaaible secondhalf 

- Column headings accross both datasets are similar but not the same eg. Order ID and Order_ID 

- Order_Date column for the secondhalf dataset needs thorough cleaning. 

- There are variety of products which affect significant price range in the data, from items priced as low as $2.99 (batteries) to as high as $1,700 (Macbook Pro Laptop).

- Products with unit prices above $99.99 should be labeled high-level products otherwise they should be basic level

- Most orders seem to involve single units of high-cost items (e.g., iPhone or Flatscreen TV), but for low-cost items like cables and batteries, customers are purchasing multiple units per order.



#### **EDA & Data Cleaning**

##### Firsthalf Data Understanding

In [47]:
# Concatenate all the DataFrames in df_firsthalf
firsthalf_combined = pd.concat(df_firsthalf, ignore_index=True)
print(firsthalf_combined.head())  


  Order ID                   Product Quantity Ordered Price Each  \
0   141234                    iPhone                1        700   
1   141235  Lightning Charging Cable                1      14.95   
2   141236          Wired Headphones                2      11.99   
3   141237          27in FHD Monitor                1     149.99   
4   141238          Wired Headphones                1      11.99   

       Order Date                       Purchase Address  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4  01/25/19 11:59          387 10th St, Austin, TX 73301  


In [48]:
#Check shape of firsthalf_combined
firsthalf_combined.shape

#Check info of firsthalf_combined 
firsthalf_combined.info()

#Check for first rows of heading
firsthalf_combined.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85625 entries, 0 to 85624
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          85380 non-null  object
 1   Product           85380 non-null  object
 2   Quantity Ordered  85380 non-null  object
 3   Price Each        85380 non-null  object
 4   Order Date        85380 non-null  object
 5   Purchase Address  85380 non-null  object
dtypes: object(6)
memory usage: 3.9+ MB


Order ID                   Product Quantity Ordered Price Each  \
0   141234                    iPhone                1        700   
1   141235  Lightning Charging Cable                1      14.95   
2   141236          Wired Headphones                2      11.99   
3   141237          27in FHD Monitor                1     149.99   
4   141238          Wired Headphones                1      11.99   

       Order Date                       Purchase Address  
0  01/22/19 21:25        944 Walnut St, Boston, MA 02215  
1  01/28/19 14:15       185 Maple St, Portland, OR 97035  
2  01/17/19 13:33  538 Adams St, San Francisco, CA 94016  
3  01/05/19 20:33     738 10th St, Los Angeles, CA 90001  
4  01/25/19 11:59          387 10th St, Austin, TX 73301

In [49]:
unique_counts = firsthalf_combined.nunique()
print(unique_counts)


Order ID            81677
Product                20
Quantity Ordered        9
Price Each             23
Order Date          66474
Purchase Address    72960
dtype: int64


In [50]:
#Describe firsthalf_combined 
firsthalf_combined.describe()

Order ID               Product Quantity Ordered Price Each  \
count      85380                 85380            85380      85380   
unique     81677                    20                9         23   
top     Order ID  USB-C Charging Cable                1      11.95   
freq         160                  9952            77205       9952   

        Order Date  Purchase Address  
count        85380             85380  
unique       66474             72960  
top     Order Date  Purchase Address  
freq           160               160

In [51]:
# Check for missing values in the first half data
print(firsthalf_combined.isnull().sum())

Order ID            245
Product             245
Quantity Ordered    245
Price Each          245
Order Date          245
Purchase Address    245
dtype: int64


#### **Insights of firsthalf_combined**
- The first half combined data has a total of  101225 and 6 columns

- The columns `Quantity Ordered`, `Price Each`, and `Order Date` will need conversion into appropriate data types (e.g., integer, float, datetime) for further analysis.

- The unique values gives a lot of insight about the data especially in dealing with duplicates in the `Order ID` column

- 245 rows are completely missing values across all columns, which indicates incomplete transactions or errors during data entry.



#### *Data Exploring and Cleaning*

In [52]:
# Drop rows with any missing values and assign the result to a new variable
firsthalf_cleaned = firsthalf_combined.dropna(how='any')

# Verify no more missing values
print(firsthalf_cleaned.isnull().sum())


Order ID            0
Product             0
Quantity Ordered    0
Price Each          0
Order Date          0
Purchase Address    0
dtype: int64


In [53]:
# Check and remove duplicates in 'Order ID' and full rows

# Check and remove duplicates in 'Order ID'
firsthalf_cleaned[firsthalf_cleaned.duplicated(subset='Order ID', keep=False)]
print(f"Duplicates in 'Order ID': {len(firsthalf_cleaned)}")

# Drop duplicates in 'Order ID', keep first occurrence
firsthalf_cleaned = firsthalf_cleaned.drop_duplicates(subset='Order ID', keep='first')

# Check and remove full-row duplicates
firsthalf_cleaned[firsthalf_cleaned.duplicated(keep=False)]
print(f"Full row duplicates: {len(firsthalf_cleaned)}")

# Drop full-row duplicates
firsthalf_cleaned = firsthalf_cleaned.drop_duplicates(keep='first')

# Recheck missing values in each column
missing_values = firsthalf_cleaned.isnull().sum()
print("\nMissing values in each column:\n", missing_values)

# Final check of data structure and types
print("\nData structure after cleanup:")
print(firsthalf_cleaned.info())


Duplicates in 'Order ID': 85380
Full row duplicates: 81677

Missing values in each column:
 Order ID            0
Product             0
Quantity Ordered    0
Price Each          0
Order Date          0
Purchase Address    0
dtype: int64

Data structure after cleanup:
<class 'pandas.core.frame.DataFrame'>
Index: 81677 entries, 0 to 85624
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Order ID          81677 non-null  object
 1   Product           81677 non-null  object
 2   Quantity Ordered  81677 non-null  object
 3   Price Each        81677 non-null  object
 4   Order Date        81677 non-null  object
 5   Purchase Address  81677 non-null  object
dtypes: object(6)
memory usage: 4.4+ MB
None


In [54]:
#Correction of datatypes

# Convert 'Order ID' to integer
firsthalf_cleaned['Order ID'] = pd.to_numeric(firsthalf_cleaned['Order ID'], errors='coerce').astype('Int64')

# Convert 'Quantity Ordered' to integer
firsthalf_cleaned['Quantity Ordered'] = pd.to_numeric(firsthalf_cleaned['Quantity Ordered'], errors='coerce').astype('Int64')

# Convert 'Price Each' to float
firsthalf_cleaned['Price Each'] = pd.to_numeric(firsthalf_cleaned['Price Each'], errors='coerce').astype(float)

# Convert 'Order Date' to datetime format
firsthalf_cleaned['Order Date'] = pd.to_datetime(firsthalf_cleaned['Order Date'], errors='coerce')

# Verify the data types after conversion
firsthalf_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 81677 entries, 0 to 85624
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Order ID          81676 non-null  Int64         
 1   Product           81677 non-null  object        
 2   Quantity Ordered  81676 non-null  Int64         
 3   Price Each        81676 non-null  float64       
 4   Order Date        81676 non-null  datetime64[ns]
 5   Purchase Address  81677 non-null  object        
dtypes: Int64(2), datetime64[ns](1), float64(1), object(2)
memory usage: 4.5+ MB


In [55]:
# Check the shape of the cleaned data
print(firsthalf_cleaned.shape)

# Print the first few rows of the cleaned dataset
print(firsthalf_cleaned.head())

(81677, 6)
   Order ID                   Product  Quantity Ordered  Price Each  \
0    141234                    iPhone                 1      700.00   
1    141235  Lightning Charging Cable                 1       14.95   
2    141236          Wired Headphones                 2       11.99   
3    141237          27in FHD Monitor                 1      149.99   
4    141238          Wired Headphones                 1       11.99   

           Order Date                       Purchase Address  
0 2019-01-22 21:25:00        944 Walnut St, Boston, MA 02215  
1 2019-01-28 14:15:00       185 Maple St, Portland, OR 97035  
2 2019-01-17 13:33:00  538 Adams St, San Francisco, CA 94016  
3 2019-01-05 20:33:00     738 10th St, Los Angeles, CA 90001  
4 2019-01-25 11:59:00          387 10th St, Austin, TX 73301  


#### **Secondhalf Data Understanding**

In [56]:
# Combine the second half of the year into one DataFrame
secondhalf_combined = pd.concat(df_secondhalf, ignore_index=True)
print(secondhalf_combined.head())

   Order_ID                   Product  Quantity_Ordered  Price_Each  \
0  222910.0  Apple Airpods Headphones               1.0      150.00   
1  222911.0             Flatscreen TV               1.0      300.00   
2  222912.0     AA Batteries (4-pack)               1.0        3.84   
3  222913.0     AA Batteries (4-pack)               1.0        3.84   
4  222914.0    AAA Batteries (4-pack)               5.0        2.99   

           Order_Date                  Purchase_Address  
0 2026-07-19 16:51:00   389 South St, Atlanta, GA 30301  
1 2005-07-19 08:55:00     590 4th St, Seattle, WA 98101  
2 2029-07-19 12:41:00    861 Hill St, Atlanta, GA 30301  
3 2028-07-19 10:15:00   190 Ridge St, Atlanta, GA 30301  
4 2031-07-19 02:13:00  824 Forest St, Seattle, WA 98101  


In [57]:
#Check shape of secondhalf_combined
secondhalf_combined.shape

#Check info of secondhalf_combined
secondhalf_combined.info()

#Check for first rows of heading
secondhalf_combined.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101225 entries, 0 to 101224
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Order_ID          100730 non-null  float64       
 1   Product           100925 non-null  object        
 2   Quantity_Ordered  100730 non-null  float64       
 3   Price_Each        100730 non-null  float64       
 4   Order_Date        100730 non-null  datetime64[ns]
 5   Purchase_Address  100925 non-null  object        
dtypes: datetime64[ns](1), float64(3), object(2)
memory usage: 4.6+ MB


Order_ID                   Product  Quantity_Ordered  Price_Each  \
0  222910.0  Apple Airpods Headphones               1.0      150.00   
1  222911.0             Flatscreen TV               1.0      300.00   
2  222912.0     AA Batteries (4-pack)               1.0        3.84   
3  222913.0     AA Batteries (4-pack)               1.0        3.84   
4  222914.0    AAA Batteries (4-pack)               5.0        2.99   

           Order_Date                  Purchase_Address  
0 2026-07-19 16:51:00   389 South St, Atlanta, GA 30301  
1 2005-07-19 08:55:00     590 4th St, Seattle, WA 98101  
2 2029-07-19 12:41:00    861 Hill St, Atlanta, GA 30301  
3 2028-07-19 10:15:00   190 Ridge St, Atlanta, GA 30301  
4 2031-07-19 02:13:00  824 Forest St, Seattle, WA 98101

In [58]:
#Describe secondhalf_combined 
secondhalf_combined.describe(include='all').T

count unique                   top   freq  \
Order_ID          100730.0    NaN                   NaN    NaN   
Product             100925     20  USB-C Charging Cable  11951   
Quantity_Ordered  100730.0    NaN                   NaN    NaN   
Price_Each        100730.0    NaN                   NaN    NaN   
Order_Date          100730    NaN                   NaN    NaN   
Purchase_Address    100925  84781      Purchase Address    195   

                                           mean                  min  \
Order_ID                          271309.400338             222910.0   
Product                                     NaN                  NaN   
Quantity_Ordered                       1.124432                  1.0   
Price_Each                           183.043312                 2.99   
Order_Date        2016-09-06 10:53:12.888712192  2001-01-20 00:10:00   
Purchase_Address                            NaN                  NaN   

                                  25%                  50%  \
Order_ID                    247141.25             271335.5   
Product                           NaN                  NaN   
Quantity_Ordered                  1.0                  1.0   
Price_Each                      11.95                14.95   
Order_Date        2008-11-19 20:50:00  2016-10-19 09:22:00   
Purchase_Address                  NaN                  NaN   

                                  75%                  max           std  
Order_ID                    295475.75             319670.0  27917.487448  
Product                           NaN                  NaN           NaN  
Quantity_Ordered                  1.0                  9.0       0.44642  
Price_Each                      150.0               1700.0    332.133484  
Order_Date        2024-07-19 13:24:00  2031-12-19 23:53:00           NaN  
Purchase_Address                  NaN                  NaN           NaN

In [59]:
# Check for missing values
print(secondhalf_combined.isnull().sum())

Order_ID            495
Product             300
Quantity_Ordered    495
Price_Each          495
Order_Date          495
Purchase_Address    300
dtype: int64


#### **Insights from Secondhalf_combined**

- The combined second half data has 101225 entries total 6 columns

- Columns Order_ID, Quantity_Ordered, Price_Each, Order_Date have 495 missing values each and 300 missing values for Product and Purchase_Address.

- Order date column must be cleaned and converted because it shows unrealistic future dates, such as 2031-12-19, indicate possible data entry errors.

- Data types of order Id and quantity must be changed to int

#### *Data Exploring and Cleaning*

In [60]:
# Drop rows'Order ID', 'Quantity Ordered', and 'Price Each'
secondhalf_cleaned = secondhalf_combined.dropna(subset=['Order_ID', 'Quantity_Ordered', 'Price_Each'])


#For other columns, you may want to fill missing data
secondhalf_cleaned['Product'].fillna('Unknown', inplace=True)
secondhalf_cleaned['Purchase_Address'].fillna('Unknown Address', inplace=True)
print(secondhalf_cleaned.head())


   Order_ID                   Product  Quantity_Ordered  Price_Each  \
0  222910.0  Apple Airpods Headphones               1.0      150.00   
1  222911.0             Flatscreen TV               1.0      300.00   
2  222912.0     AA Batteries (4-pack)               1.0        3.84   
3  222913.0     AA Batteries (4-pack)               1.0        3.84   
4  222914.0    AAA Batteries (4-pack)               5.0        2.99   

           Order_Date                  Purchase_Address  
0 2026-07-19 16:51:00   389 South St, Atlanta, GA 30301  
1 2005-07-19 08:55:00     590 4th St, Seattle, WA 98101  
2 2029-07-19 12:41:00    861 Hill St, Atlanta, GA 30301  
3 2028-07-19 10:15:00   190 Ridge St, Atlanta, GA 30301  
4 2031-07-19 02:13:00  824 Forest St, Seattle, WA 98101  


In [61]:
# Check for duplicates
print("Duplicate rows:")
print(secondhalf_cleaned[secondhalf_cleaned.duplicated()])
print(f"Number of duplicate rows: {secondhalf_cleaned.shape[0]}")

# Drop duplicates based on 'Order_ID' and keep the first occurrence
secondhalf_cleaned = secondhalf_cleaned.drop_duplicates(subset=['Order_ID'], keep='first')
print("\nDataFrame after dropping duplicates:")
print(secondhalf_cleaned.head())

# Verify the removal of duplicates
print("\nNumber of remaining duplicates:", secondhalf_cleaned.duplicated().sum())

Duplicate rows:
       Order_ID                     Product  Quantity_Ordered  Price_Each  \
2880   225674.0  Bose SoundSport Headphones               1.0   99.989998   
3580   226333.0      AAA Batteries (4-pack)               1.0    2.990000   
3674   226421.0                Google Phone               1.0  600.000000   
3784   226526.0      AAA Batteries (4-pack)               1.0    2.990000   
3887   226625.0            Wired Headphones               1.0   11.990000   
...         ...                         ...               ...         ...   
96002  314675.0       AA Batteries (4-pack)               1.0    3.840000   
96560  315204.0            Wired Headphones               1.0   11.990000   
97348  315955.0             ThinkPad Laptop               1.0  999.989990   
97575  316173.0      AAA Batteries (4-pack)               1.0    2.990000   
99460  317971.0       AA Batteries (4-pack)               1.0    3.840000   

               Order_Date                          Purchase

In [62]:
# Convert 'Order_ID' to integer
secondhalf_cleaned['Order_ID'] = secondhalf_cleaned['Order_ID'].astype(int)

# Convert 'Quantity_Ordered' to integer
secondhalf_cleaned['Quantity_Ordered'] = secondhalf_cleaned['Quantity_Ordered'].astype(int)

# Convert 'Order_Date' to datetime
secondhalf_cleaned['Order_Date'] = pd.to_datetime(secondhalf_cleaned['Order_Date'], errors='coerce')

# Verify the data types after conversion
print(secondhalf_cleaned.info())


<class 'pandas.core.frame.DataFrame'>
Index: 96761 entries, 0 to 101224
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Order_ID          96761 non-null  int64         
 1   Product           96761 non-null  object        
 2   Quantity_Ordered  96761 non-null  int64         
 3   Price_Each        96761 non-null  float64       
 4   Order_Date        96761 non-null  datetime64[ns]
 5   Purchase_Address  96761 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 5.2+ MB
None


#### **Merging of both Data sets (firsthalf_cleaned and secondhalf_cleaned)**

In [63]:
# For the first half (January to June), ensuring column names are consistent
firsthalf_cleaned.rename(columns={
    'Order ID': 'Order_ID',
    'Quantity Ordered': 'Quantity_Ordered',
    'Price Each': 'Price_Each',
    'Order Date': 'Order_Date',
    'Purchase Address': 'Purchase_Address'
}, inplace=True)

# Now concatenate the datasets
df_combined = pd.concat([firsthalf_cleaned, secondhalf_cleaned], ignore_index=True)

# Show the first few rows to confirm the data is concatenated properly
print("Combined DataFrame:")
print(df_combined.head())


Combined DataFrame:
   Order_ID                   Product  Quantity_Ordered  Price_Each  \
0    141234                    iPhone                 1      700.00   
1    141235  Lightning Charging Cable                 1       14.95   
2    141236          Wired Headphones                 2       11.99   
3    141237          27in FHD Monitor                 1      149.99   
4    141238          Wired Headphones                 1       11.99   

           Order_Date                       Purchase_Address  
0 2019-01-22 21:25:00        944 Walnut St, Boston, MA 02215  
1 2019-01-28 14:15:00       185 Maple St, Portland, OR 97035  
2 2019-01-17 13:33:00  538 Adams St, San Francisco, CA 94016  
3 2019-01-05 20:33:00     738 10th St, Los Angeles, CA 90001  
4 2019-01-25 11:59:00          387 10th St, Austin, TX 73301  


In [64]:
df_combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178438 entries, 0 to 178437
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Order_ID          178437 non-null  Int64         
 1   Product           178438 non-null  object        
 2   Quantity_Ordered  178437 non-null  Int64         
 3   Price_Each        178437 non-null  float64       
 4   Order_Date        178437 non-null  datetime64[ns]
 5   Purchase_Address  178438 non-null  object        
dtypes: Int64(2), datetime64[ns](1), float64(1), object(2)
memory usage: 8.5+ MB


#### *Further Exploring and Cleaning*

In [65]:
# Loop through each column and print unique values
for column in df_combined.columns:
    unique_values = df[column].unique()
    print(f"Unique values in column '{column}':")
    print(unique_values)
    print()  # Blank line for better readability

Unique values in column 'Order_ID':
[295665. 295666. 295667. ... 319668. 319669. 319670.]

Unique values in column 'Product':
['Macbook Pro Laptop' 'LG Washing Machine' 'USB-C Charging Cable'
 '27in FHD Monitor' 'AA Batteries (4-pack)' 'Bose SoundSport Headphones'
 'AAA Batteries (4-pack)' 'ThinkPad Laptop' 'Lightning Charging Cable'
 'Google Phone' 'Wired Headphones' 'Apple Airpods Headphones'
 'Vareebadd Phone' 'iPhone' '20in Monitor' '34in Ultrawide Monitor'
 'Flatscreen TV' '27in 4K Gaming Monitor' 'Product' None 'LG Dryer']

Unique values in column 'Quantity_Ordered':
[ 1.  2.  4.  3. nan  7.  5.  6.]

Unique values in column 'Price_Each':
[1700.          600.           11.94999981  149.99000549    3.83999991
   99.98999786    2.99000001  999.98999023   14.94999981   11.98999977
  150.          400.          700.          109.98999786  379.98999023
  300.          389.98999023           nan]

Unique values in column 'Order_Date':
<DatetimeArray>
['2030-12-19 00:01:00', '2029-12-19

In [66]:
# Check for missing values in critical columns
print(df_combined.isnull().sum())

Order_ID            1
Product             0
Quantity_Ordered    1
Price_Each          1
Order_Date          1
Purchase_Address    0
dtype: int64


In [67]:
# Check for duplicates based on Order_ID
#duplicates = df_combined[df_combined.duplicated(subset='Order_ID')]
#print(f"Number of duplicates: {duplicates.shape[0]}")

# If duplicates exist, drop them
#df_combined.drop_duplicates(subset='Order_ID', inplace=True)


In [68]:
# Check for missing values in critical columns
print(df_combined.isnull().sum())

# Drop rows where critical columns are null
#df_combined.dropna(subset=['Order_ID', 'Quantity_Ordered', 'Price_Each', 'Order_Date'], inplace=True)


Order_ID            1
Product             0
Quantity_Ordered    1
Price_Each          1
Order_Date          1
Purchase_Address    0
dtype: int64


In [69]:
# Check for any NaT values (invalid dates)
if df_combined['Order_Date'].isna().any():
    print("There are invalid dates in the 'Order Date' column.")

There are invalid dates in the 'Order Date' column.


##### *Order Id Column clean*

In [70]:
# Using pandas' unique() function
unique_values = df['Order_ID'].unique()
print(unique_values)


# Count the number of duplicated Order_IDs
num_duplicates = df['Order_ID'].duplicated().sum()

print(f"Number of duplicated Order_IDs: {num_duplicates}")


# Drop duplicate Order_ID rows, keeping the first occurrence
df_unique = df.drop_duplicates(subset='Order_ID', keep='first')

# Verify the result
print("DataFrame after removing duplicate Order_ID rows:")
print(df_unique)


[295665. 295666. 295667. ... 319668. 319669. 319670.]
Number of duplicated Order_IDs: 1110
DataFrame after removing duplicate Order_ID rows:
       Order_ID                     Product  Quantity_Ordered   Price_Each  \
0      295665.0          Macbook Pro Laptop               1.0  1700.000000   
1      295666.0          LG Washing Machine               1.0   600.000000   
2      295667.0        USB-C Charging Cable               1.0    11.950000   
3      295668.0            27in FHD Monitor               1.0   149.990005   
4      295669.0        USB-C Charging Cable               1.0    11.950000   
...         ...                         ...               ...          ...   
25112  319666.0    Lightning Charging Cable               1.0    14.950000   
25113  319667.0       AA Batteries (4-pack)               2.0     3.840000   
25114  319668.0             Vareebadd Phone               1.0   400.000000   
25115  319669.0            Wired Headphones               1.0    11.990000   
2

In [71]:
Products
# Using pandas' unique() function
unique_values = df['Product'].unique()
print(unique_values)

NameError: name 'Products' is not defined

In [ ]:
qty Ordered 

# Step 1: Convert 'Quantity_Ordered' to numeric, coercing non-numeric values to NaN
df['Quantity_Ordered'] = pd.to_numeric(df['Quantity_Ordered'], errors='coerce')

# Step 2: Drop NaN values that result from invalid conversions
df = df.dropna(subset=['Quantity_Ordered'])

# Step 3: Convert the column to integers
df['Quantity_Ordered'] = df['Quantity_Ordered'].astype(int)

# Step 4: Get the unique values
unique_values = df['Quantity_Ordered'].unique()

# Step 5: Print the unique integer values
print(unique_values)

In [ ]:
price each
# Using pandas' unique() function
unique_values = df['Price_Each'].unique()
print(unique_values)


# Clean the 'Price_Each' column
df['Price_Each'] = pd.to_numeric(df['Price_Each'], errors='coerce')  # Convert to numeric, forcing errors to NaN

# Drop NaN values
df_cleaned = df.dropna(subset=['Price_Each'])

# Round to 2 decimal places
df_cleaned['Price_Each'] = df_cleaned['Price_Each'].round(2)

# Check the unique values after cleaning
unique_values_cleaned = df_cleaned['Price_Each'].unique()
print(unique_values_cleaned)

# Round values in the 'Price_Each' column to 2 decimal places
df['Price_Each'] = df['Price_Each'].round(2)

# Verify the result
print(df['Price_Each'].head())



# Ensure that 'Quantity_Ordered' and 'Price_Each' are numeric
df_unique['Quantity_Ordered'] = pd.to_numeric(df_unique['Quantity_Ordered'], errors='coerce')
df['Price_Each'] = pd.to_numeric(df_unique['Price_Each'], errors='coerce')

# Create a new column 'Sales' by multiplying 'Quantity_Ordered' and 'Price_Each'
df_unique['Sales'] = df_unique['Quantity_Ordered'] * df_unique['Price_Each']

# Verify the result
print(df_unique[['Quantity_Ordered', 'Price_Each', 'Sales']].head())









In [ ]:
order date 
# Convert 'Order Date' to datetime and handle errors
df_combined['Order_Date'] = pd.to_datetime(df_combined['Order_Date'], format='%m/%d/%y %H:%M', errors='coerce')

# Check for any NaT values (invalid dates)
if df_combined['Order_Date'].isna().any():
    print("There are invalid dates in the 'Order_Date' column.")

# Optionally, handle invalid dates (e.g., fill with a default date or leave as NaT)
df_combined['Order_Date'] = df_combined['Order_Date'].fillna(pd.Timestamp('1970-01-01 00:00:00'))  # Example: fill with a default date

# Format only the valid dates
df_combined['Order_Date'] = df_combined['Order_Date'].apply(lambda x: x.strftime('%Y-%m-%d %H:%M:%S.%f') if pd.notna(x) else None)

# Display the DataFrame
print(df_combined)


# Ensure that Order_Date is in datetime format
df_unique['Order_Date'] = pd.to_datetime(df_unique['Order_Date'])

# Create a new column 'Month' by extracting the month from 'Order_Date'
df_unique['Month'] = df_unique['Order_Date'].dt.month

# Verify the result
print(df_unique[['Order_Date', 'Month']].head())



In [ ]:
purchase address 

# Using pandas' unique() function
unique_values = df['Purchase_Address'].unique()
print(unique_values)


# Split the Purchase_Address column and extract the City

# Define a function to extract city from the address
def get_city(address):
    return address.split(",")[1].strip()

# Apply the function to create a new 'City' column
df_unique['City'] = df_unique['Purchase_Address'].apply(get_city)

# Display the updated DataFrame with the new 'City' column
print(df_unique.head())



#### **General Insights**
-

### **Data Preparation** 

### **Modelling and Evaluation**

### **Deployment**